types.go

```diff
@@ -0,0 +1,79 @@
+package fq
+
+import (
+	"math"
+)
+
+// TODO(aaron-prindle) currently testing with one concurrent request
+const C = 1 // const C = 300
+
+// TODO(aaron-prindle) currently service time "G" is not implemented entirely
+const G = 100000 //   100000 nanoseconds = 100 milliseconds || const G = 60000000 nanoseconds = 1 minute
+
+type Packet struct {
+	// request   http.Request
+	item interface{}
+	// virfinish float64
+	size int
+	// key       uint64
+	queueidx  int
+	seq       int
+	starttime float64
+}
+
+type Queue struct {
+	Packets           []*Packet
+	virstart          float64
+	RequestsExecuting int
+	// RequestsExecuting []*Packet
+}
+
+func (q *Queue) enqueue(packet *Packet) {
+	q.Packets = append(q.Packets, packet)
+}
+
+func (q *Queue) dequeue() (*Packet, bool) {
+	if len(q.Packets) == 0 {
+		return nil, false
+	}
+	packet := q.Packets[0]
+	q.Packets = q.Packets[1:]
+	return packet, true
+}
+
+func initQueues(n int, key uint64) []*Queue {
+	queues := []*Queue{}
+	for i := 0; i < n; i++ {
+		queues = append(queues, &Queue{
+			Packets: []*Packet{},
+		})
+	}
+	return queues
+}
+
+func (p *Packet) virfinish(J int, q *FQScheduler) float64 {
+	// The virtual finish time of request number J in the queue
+	// (counting from J=1 for the head) is J * G + (virtual start time).
+
+	J++ // counting from J=1 for the head (eg: queue.Packets[0] -> J=1)
+	// if J*G overflows
+	//    throw an error?
+	jg := float64(J * G)
+	if math.IsInf(jg, 1) {
+		panic("float overflowed")
+	}
+	return jg + q.queues[p.queueidx].virstart
+
+}
+
+func (p *Packet) finishRequest(q *FQScheduler) {
```

Comment:
Rather than a private method of Packet, this should be a public entry method of FQScheduler.  The simplest API would combine this with Dequeue, since it is when one packet is finished that it is time to pick the next one to serve.  It would be OK to leave the current split; in that case both entry methods would need to call syncTime.

Reply:
I have moved this to be a public entry method of FQScheduler that takes a Packet.  I have added a currently unused method, FinishPacketAndDeque and which combines the two.  I understand how they could be used jointly but I think we would still want each seperate method as well.